scipy/interpolate/_rgi_cython.pyx

```diff
@@ -0,0 +1,181 @@
+            if xval < x[mid]:
+                # mid < high
+                high = mid
+            elif xval >= x[mid + 1]:
+                low = mid + 1
+            else:
+                # x[mid] <= xval < x[mid+1]
+                low = mid
+                break
+
+        interval = low
+
+    return interval
+
+
+def evaluate_linear(values, indices, norm_distances, out_of_bounds):
+    # slice for broadcasting over trailing dimensions in self.values
+    vslice = (slice(None),) + (None,) * (values.ndim - len(indices))
+
+    # Compute shifting up front before zipping everything together
+    shift_norm_distances = [1 - yi for yi in norm_distances]
+    shift_indices = [i + 1 for i in indices]
+
+    # The formula for linear interpolation in 2d takes the form:
+    # values = self.values[(i0, i1)] * (1 - y0) * (1 - y1) + \
+    #          self.values[(i0, i1 + 1)] * (1 - y0) * y1 + \
+    #          self.values[(i0 + 1, i1)] * y0 * (1 - y1) + \
+    #          self.values[(i0 + 1, i1 + 1)] * y0 * y1
+    # We pair i with 1 - yi (zipped1) and i + 1 with yi (zipped2)
+    zipped1 = zip(indices, shift_norm_distances)
+    zipped2 = zip(shift_indices, norm_distances)
+
+    # Take all products of zipped1 and zipped2 and iterate over them
+    # to get the terms in the above formula. This corresponds to iterating
+    # over the vertices of a hypercube.
+    hypercube = itertools.product(*zip(zipped1, zipped2))
+    value = np.array([0.])
+    for h in hypercube:
+        edge_indices, weights = zip(*h)
+        weight = np.array([1.])
+        for w in weights:
+            weight = weight * w
+        value = value + np.asarray(values[edge_indices]) * weight[vslice]
+    return value
+
+
+@cython.wraparound(False)
+@cython.boundscheck(True)
+@cython.cdivision(True)
+def find_indices(tuple grid not None, double[:, :] xi):
+    cdef long i, j, I, J, grid_i_size
+    cdef double denom, value
+    cdef long[:,::1] indices
+    cdef double[:,::1] norm_distances
+    cdef double[::1] grid_i
+
+    cdef int index = 0
+    # find relevant edges between which xi are situated
+    indices = np.empty_like(xi, dtype=int)
+    # compute distance to lower edge in unity units
+    norm_distances = np.zeros_like(xi, dtype=float)
+
+    # indexes to iterate over
+    I = xi.shape[0]
+    J = xi.shape[1]
+
+    # iterate through dimensions
+    for i in range(I):
+        grid_i = grid[i]
+        grid_i_size = grid_i.shape[0]
+
+        if grid_i_size == 1:
+            # special case length-one dimensions
+            for j in range(J):
+                indices[i, j] = -1    # Should equal 0. Setting it to -1 is a hack: evaluate_linear looks at indices [i, i+1]
+                                      # which both end up =0 with wraparound. 
+                                      # Conclusion: change -1 to 0 here together with refactoring evaluate_linear, which
+                                      # will also need to special-case length-one axes
```

Comment:
Nitpick: this prevent potential large diff in case of a small change. This also reads better to some. 
```suggestion
                # Should equal 0. Setting it to -1 is a hack: evaluate_linear 
                # looks at indices [i, i+1] which both end up =0 with wraparound. 
                # Conclusion: change -1 to 0 here together with refactoring
                # evaluate_linear, which will also need to special-case
                # length-one axes
                indices[i, j] = -1
```